File: bots/minigames/cogs/tictactoe.py
"""
bots/minigames/cogs/tictactoe.py
──────────────────────────────────
TicTacToe – Player vs Bot.
Difficulty levels: Easy (random), Medium (logic), Hard (Minimax).
"""

from __future__ import annotations

import asyncio
import logging
import random
from typing import Optional

import discord
from discord import app_commands
from discord.ext import commands
from discord.ui import Button, View

from core.points_manager import add_points, get_pts, notify_rewards, points_footer

log = logging.getLogger(__name__)

EMPTY  = ""
PLAYER = "X"
BOT    = "O"

WINNING_COMBOS = [
    (0, 1, 2), (3, 4, 5), (6, 7, 8),
    (0, 3, 6), (1, 4, 7), (2, 5, 8),
    (0, 4, 8), (2, 4, 6),
]

DIFFICULTY_LABELS = {
    "easy":   "🟢 Easy",
    "medium": "🟡 Medium",
    "hard":   "🔴 Hard",
}

CELL_EMOJIS = {PLAYER: "❌", BOT: "⭕", EMPTY: "⬜"}


# ─── Game logic ───────────────────────────────────────────────────────────────

def check_winner(board: list) -> Optional[str]:
    for a, b, c in WINNING_COMBOS:
        if board[a] and board[a] == board[b] == board[c]:
            return board[a]
    return None


def is_draw(board: list) -> bool:
    return all(c != EMPTY for c in board) and check_winner(board) is None


def available_moves(board: list) -> list:
    return [i for i, c in enumerate(board) if c == EMPTY]


def _ai_easy(board: list) -> int:
    return random.choice(available_moves(board))


def _winning_move(board: list, symbol: str) -> Optional[int]:
    for move in available_moves(board):
        board[move] = symbol
        won = check_winner(board) == symbol
        board[move] = EMPTY
        if won:
            return move
    return None


def _ai_medium(board: list) -> int:
    if (m := _winning_move(board, BOT)) is not None:
        return m
    if (m := _winning_move(board, PLAYER)) is not None and random.random() > 0.25:
        return m
    if random.random() < 0.40:
        return random.choice(available_moves(board))
    for pos in [4, 0, 2, 6, 8, 1, 3, 5, 7]:
        if board[pos] == EMPTY:
            return pos
    return random.choice(available_moves(board))


def _minimax(board: list, maximizing: bool, depth: int = 0) -> int:
    winner = check_winner(board)
    if winner == BOT:    return 10 - depth
    if winner == PLAYER: return depth - 10
    if is_draw(board):   return 0
    if maximizing:
        best = -100
        for move in available_moves(board):
            board[move] = BOT
            best = max(best, _minimax(board, False, depth + 1))
            board[move] = EMPTY
        return best
    else:
        best = 100
        for move in available_moves(board):
            board[move] = PLAYER
            best = min(best, _minimax(board, True, depth + 1))
            board[move] = EMPTY
        return best


def _ai_hard(board: list) -> int:
    best_score, best_move = -100, -1
    for move in available_moves(board):
        board[move] = BOT
        score = _minimax(board, False)
        board[move] = EMPTY
        if score > best_score:
            best_score, best_move = score, move
    return best_move


# ─── Views ────────────────────────────────────────────────────────────────────

class DifficultyView(View):
    def __init__(self, player: discord.User) -> None:
        super().__init__(timeout=60)
        self.player = player
        for key, label in DIFFICULTY_LABELS.items():
            btn = Button(label=label, custom_id=f"ttt_diff_{key}", style=discord.ButtonStyle.primary)
            btn.callback = self._make_callback(key)
            self.add_item(btn)

    def _make_callback(self, difficulty: str):
        async def callback(interaction: discord.Interaction) -> None:
            if interaction.user.id != self.player.id:
                await interaction.response.send_message("❌ This is not your game!", ephemeral=True)
                return
            self.stop()
            await interaction.response.defer()
            view = TicTacToeView(self.player, difficulty)
            await interaction.edit_original_response(embed=view.build_embed(), view=view)
        return callback

    async def on_timeout(self) -> None:
        for item in self.children:
            item.disabled = True


class TicTacToeView(View):
    def __init__(self, player: discord.User, difficulty: str) -> None:
        super().__init__(timeout=120)
        self.player     = player
        self.difficulty = difficulty
        self.board      = [EMPTY] * 9
        self.game_over  = False
        self._rebuild()

    def _rebuild(self) -> None:
        self.clear_items()
        for i in range(9):
            cell  = self.board[i]
            style = (
                discord.ButtonStyle.danger    if cell == PLAYER else
                discord.ButtonStyle.primary   if cell == BOT    else
                discord.ButtonStyle.secondary
            )
            btn = Button(label=CELL_EMOJIS[cell], style=style,
                         custom_id=f"ttt_cell_{i}", row=i // 3,
                         disabled=self.game_over or cell != EMPTY)
            btn.callback = self._make_cb(i)
            self.add_item(btn)

    def _make_cb(self, index: int):
        async def callback(interaction: discord.Interaction) -> None:
            if interaction.user.id != self.player.id:
                await interaction.response.send_message("❌ This is not your game!", ephemeral=True)
                return
            if self.game_over or self.board[index] != EMPTY:
                await interaction.response.defer()
                return
            self.board[index] = PLAYER
            result = self._eval()
            if result is None:
                await asyncio.sleep(0.4)
                move = (_ai_easy if self.difficulty == "easy" else
                        _ai_medium if self.difficulty == "medium" else _ai_hard)(self.board)
                self.board[move] = BOT
                result = self._eval()
            self._rebuild()

            # Punkte vergeben wenn Spiel beendet
            pts_delta, old, new = 0, 0, 0
            if result is not None:
                outcome = "win" if result == PLAYER else "lose" if result == BOT else "draw"
                pts_delta = get_pts("tictactoe", self.difficulty, outcome)
                old, new  = await add_points(interaction.user.id, pts_delta)

            await interaction.response.edit_message(
                embed=self.build_embed(result, pts_delta, new), view=self
            )
            if result is not None:
                await notify_rewards(interaction, old, new)
        return callback

    def _eval(self) -> Optional[str]:
        w = check_winner(self.board)
        if w:
            self.game_over = True
            return w
        if is_draw(self.board):
            self.game_over = True
            return "draw"
        return None

    def build_embed(self, result: Optional[str] = None, pts_delta: int = 0, total: int = 0) -> discord.Embed:
        diff   = DIFFICULTY_LABELS[self.difficulty]
        footer = "TicTacToe  •  /tictactoe to play again"
        if result is not None:
            footer += f"  •  {points_footer(pts_delta, total)}"
        if result is None:
            title, desc, color = "🎮 TicTacToe", f"Player: {self.player.mention} ❌ vs ⭕ Bot\nDifficulty: {diff}\n\nYour turn!", discord.Color.blurple()
        elif result == PLAYER:
            title, desc, color = "🏆 You won!", f"Well played, {self.player.mention}! ❌\nDifficulty: {diff}", discord.Color.green()
        elif result == BOT:
            title, desc, color = "🤖 Bot wins!", f"Better luck next time, {self.player.mention}.\nDifficulty: {diff}", discord.Color.red()
        else:
            title, desc, color = "🤝 Draw!", f"No winner this time, {self.player.mention}.\nDifficulty: {diff}", discord.Color.gold()
        embed = discord.Embed(title=title, description=desc, color=color)
        embed.set_footer(text=footer)
        return embed

    async def on_timeout(self) -> None:
        self.game_over = True
        for item in self.children:
            item.disabled = True


# ─── Cog ──────────────────────────────────────────────────────────────────────

class TicTacToeCog(commands.Cog, name="TicTacToe"):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @app_commands.command(name="tictactoe", description="Play TicTacToe against the bot!")
    async def tictactoe(self, interaction: discord.Interaction) -> None:
        embed = discord.Embed(
            title="🎮 TicTacToe – Choose Difficulty",
            description=f"Hello {interaction.user.mention}! You play as ❌.\n\nChoose a difficulty:",
            color=discord.Color.blurple(),
        )
        await interaction.response.send_message(embed=embed, view=DifficultyView(interaction.user))


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(TicTacToeCog(bot))

File: bots/minigames/commands/tictactoe.js
const { SlashCommandBuilder, ButtonBuilder, ButtonStyle, ActionRowBuilder, EmbedBuilder } = require('discord.js');
const { addPoints, getPts, notifyRewards, pointsFooter } = require('../../../core/pointsManager');

const EMPTY  = '';
const PLAYER = 'X';
const BOT    = 'O';

const WINNING_COMBOS = [
  [0,1,2],[3,4,5],[6,7,8],
  [0,3,6],[1,4,7],[2,5,8],
  [0,4,8],[2,4,6],
];

const CELL_EMOJIS = { [PLAYER]: '❌', [BOT]: '⭕', [EMPTY]: '⬜' };
const DIFF_LABELS  = { easy: '🟢 Easy', medium: '🟡 Medium', hard: '🔴 Hard' };

function checkWinner(board) {
  for (const [a, b, c] of WINNING_COMBOS) {
    if (board[a] && board[a] === board[b] && board[b] === board[c]) return board[a];
  }
  return null;
}

function isDraw(board) { return board.every(c => c !== EMPTY) && !checkWinner(board); }
function available(board) { return board.map((c, i) => c === EMPTY ? i : -1).filter(i => i !== -1); }

function aiEasy(board) { const m = available(board); return m[Math.floor(Math.random() * m.length)]; }

function winningMove(board, sym) {
  for (const move of available(board)) {
    board[move] = sym;
    const won = checkWinner(board) === sym;
    board[move] = EMPTY;
    if (won) return move;
  }
  return null;
}

function aiMedium(board) {
  let m;
  if ((m = winningMove(board, BOT))    !== null) return m;
  if ((m = winningMove(board, PLAYER)) !== null && Math.random() > 0.25) return m;
  if (Math.random() < 0.4) return aiEasy(board);
  for (const pos of [4, 0, 2, 6, 8, 1, 3, 5, 7]) {
    if (board[pos] === EMPTY) return pos;
  }
  return aiEasy(board);
}

function minimax(board, maximizing, depth = 0) {
  const w = checkWinner(board);
  if (w === BOT)    return 10 - depth;
  if (w === PLAYER) return depth - 10;
  if (isDraw(board)) return 0;
  if (maximizing) {
    let best = -100;
    for (const move of available(board)) {
      board[move] = BOT;
      best = Math.max(best, minimax(board, false, depth + 1));
      board[move] = EMPTY;
    }
    return best;
  } else {
    let best = 100;
    for (const move of available(board)) {
      board[move] = PLAYER;
      best = Math.min(best, minimax(board, true, depth + 1));
      board[move] = EMPTY;
    }
    return best;
  }
}

function aiHard(board) {
  let bestScore = -100, bestMove = -1;
  for (const move of available(board)) {
    board[move] = BOT;
    const score = minimax(board, false);
    board[move] = EMPTY;
    if (score > bestScore) { bestScore = score; bestMove = move; }
  }
  return bestMove;
}

function buildBoard(board, difficulty, gameOver) {
  const rows = [];
  for (let r = 0; r < 3; r++) {
    const row = new ActionRowBuilder();
    for (let c = 0; c < 3; c++) {
      const i    = r * 3 + c;
      const cell = board[i];
      const style = cell === PLAYER ? ButtonStyle.Danger
                  : cell === BOT    ? ButtonStyle.Primary
                  : ButtonStyle.Secondary;
      row.addComponents(
        new ButtonBuilder()
          .setCustomId(`ttt_cell_${i}`)
          .setLabel(CELL_EMOJIS[cell])
          .setStyle(style)
          .setDisabled(gameOver || cell !== EMPTY)
      );
    }
    rows.push(row);
  }
  return rows;
}

function buildEmbed(player, difficulty, result = null, ptsDelta = 0, total = 0) {
  const diff   = DIFF_LABELS[difficulty];
  let footer   = 'TicTacToe  •  /tictactoe to play again';
  if (result !== null) footer += `  •  ${pointsFooter(ptsDelta, total)}`;

  let title, desc, color;
  if (result === null) {
    title = '🎮 TicTacToe';
    desc  = `Player: ${player} ❌ vs ⭕ Bot\nDifficulty: ${diff}\n\nYour turn!`;
    color = 0x5865F2;
  } else if (result === PLAYER) {
    title = '🏆 You won!';
    desc  = `Well played, ${player}! ❌\nDifficulty: ${diff}`;
    color = 0x57F287;
  } else if (result === BOT) {
    title = '🤖 Bot wins!';
    desc  = `Better luck next time, ${player}.\nDifficulty: ${diff}`;
    color = 0xED4245;
  } else {
    title = '🤝 Draw!';
    desc  = `No winner this time, ${player}.\nDifficulty: ${diff}`;
    color = 0xFEE75C;
  }
  return new EmbedBuilder().setTitle(title).setDescription(desc).setColor(color).setFooter({ text: footer });
}

module.exports = {
  data: new SlashCommandBuilder().setName('tictactoe').setDescription('Play TicTacToe against the bot!'),

  async execute(interaction) {
    const diffRow = new ActionRowBuilder().addComponents(
      new ButtonBuilder().setCustomId('ttt_diff_easy').setLabel('🟢 Easy').setStyle(ButtonStyle.Success),
      new ButtonBuilder().setCustomId('ttt_diff_medium').setLabel('🟡 Medium').setStyle(ButtonStyle.Primary),
      new ButtonBuilder().setCustomId('ttt_diff_hard').setLabel('🔴 Hard').setStyle(ButtonStyle.Danger),
    );

    const startEmbed = new EmbedBuilder()
      .setTitle('🎮 TicTacToe – Choose Difficulty')
      .setDescription(`Hello ${interaction.user}! You play as ❌.\n\nChoose a difficulty:`)
      .setColor(0x5865F2);

    await interaction.reply({ embeds: [startEmbed], components: [diffRow] });
    const reply = await interaction.fetchReply();

    const diffCollector = reply.createMessageComponentCollector({
      filter: i => i.user.id === interaction.user.id && i.customId.startsWith('ttt_diff_'),
      time: 60_000, max: 1,
    });

    diffCollector.on('collect', async diffInter => {
      const difficulty = diffInter.customId.replace('ttt_diff_', '');
      const board = Array(9).fill(EMPTY);
      let gameOver = false;

      await diffInter.update({
        embeds: [buildEmbed(interaction.user, difficulty)],
        components: buildBoard(board, difficulty, false),
      });

      const gameCollector = reply.createMessageComponentCollector({
        filter: i => i.user.id === interaction.user.id && i.customId.startsWith('ttt_cell_'),
        time: 120_000,
      });

      gameCollector.on('collect', async i => {
        if (gameOver) { await i.deferUpdate(); return; }

        const idx = parseInt(i.customId.replace('ttt_cell_', ''));
        if (board[idx] !== EMPTY) { await i.deferUpdate(); return; }

        board[idx] = PLAYER;
        let result = checkWinner(board) ? PLAYER : isDraw(board) ? 'draw' : null;

        if (result === null) {
          const aiMove = difficulty === 'easy' ? aiEasy(board) : difficulty === 'medium' ? aiMedium(board) : aiHard(board);
          board[aiMove] = BOT;
          result = checkWinner(board) ? BOT : isDraw(board) ? 'draw' : null;
        }

        if (result !== null) {
          gameOver = true;
          gameCollector.stop();
        }

        let ptsDelta = 0, oldPts = 0, newPts = 0;
        if (result !== null) {
          const outcome = result === PLAYER ? 'win' : result === BOT ? 'lose' : 'draw';
          ptsDelta = getPts('tictactoe', difficulty, outcome);
          const pts = addPoints(interaction.user.id, ptsDelta);
          oldPts = pts.old; newPts = pts.new;
        }

        await i.update({
          embeds: [buildEmbed(interaction.user, difficulty, result, ptsDelta, newPts)],
          components: buildBoard(board, difficulty, gameOver),
        });

        if (result !== null) await notifyRewards(i, oldPts, newPts);
      });

      gameCollector.on('end', () => {
        if (!gameOver) {
          interaction.editReply({ components: buildBoard(board, difficulty, true) }).catch(() => {});
        }
      });
    });

    diffCollector.on('end', (collected) => {
      if (!collected.size) {
        for (const btn of diffRow.components) btn.setDisabled(true);
        interaction.editReply({ components: [diffRow] }).catch(() => {});
      }
    });
  },
};
